source/common/http/conn_pool_grid.cc

```diff
@@ -334,5 +334,58 @@ void ConnectivityGrid::onDrainReceived() {
   }
 }
 
+bool ConnectivityGrid::shouldAttemptHttp3() {
+  if (http3_status_tracker_.isHttp3Broken()) {
+    ENVOY_LOG(trace, "HTTP/3 is broken to host '{}', skipping.", host_->hostname());
+    return false;
+  }
+  if (!alternate_protocols_.has_value()) {
+    ENVOY_LOG(trace, "No alternate protocols cache. Attempting HTTP/3 to host '{}'.",
+              host_->hostname());
+    return true;
+  }
+  if (host_->address()->type() != Network::Address::Type::Ip) {
+    ENVOY_LOG(error, "Address is not an IP address");
+    ASSERT(false);
+    return false;
+  }
+  uint32_t port = host_->address()->ip()->port();
+  // TODO(RyanTheOptimist): Figure out how scheme gets plumbed in here.
+  AlternateProtocolsCache::Origin origin("https", host_->hostname(), port);
+  OptRef<const std::vector<AlternateProtocolsCache::AlternateProtocol>> protocols =
+      alternate_protocols_->findAlternatives(origin);
+  if (!protocols.has_value()) {
+    ENVOY_LOG(trace, "No alternate protocols available for host '{}', skipping HTTP/3.",
+              host_->hostname());
+    return false;
+  }
+
+  for (const AlternateProtocolsCache::AlternateProtocol& protocol : protocols.ref()) {
+    // TODO(RyanTheOptimist): Handle alternate protocols which change hostname or port.
+    if (!protocol.hostname_.empty() || protocol.port_ != port) {
+      ENVOY_LOG(trace,
+                "Alternate protocol for host '{}' attempts to change host or port, skipping.",
+                host_->hostname());
+      continue;
+    }
+
+    // TODO(RyanTheOptimist): Cache this mapping, but handle the supported versions list
+    // changing dynamically.
+    for (const quic::ParsedQuicVersion& version : quic::CurrentSupportedVersions()) {
+      if (quic::AlpnForVersion(version) == protocol.alpn_) {
+        // TODO(RyanTheOptimist): Pass this version down to the HTTP/3 pool.
+        ENVOY_LOG(trace, "HTTP/3 advertised for host '{}'", host_->hostname());
+        return true;
+      }
+    }
+
+    ENVOY_LOG(trace, "Alternate protocol for host '{}' has unsupported ALPN '{}', skipping.",
+              host_->hostname(), protocol.alpn_);
+  }
+
+  ENVOY_LOG(trace, "HTTP/3 is not available to host '{}', skipping.", host_->hostname());
+  return false;
```

Comment:
seems like a good place to start adding stats, WDYT?  Can defer to a follow up but let's make sure we have stats as we wire this all up.